File: _wpf/wpf-business-programmer/3-wpf-business-programmer-m3-exercise-files/Samples-2012/VB/WPFTemplateApp/WPFTemplateApp/ViewModels/ProductViewModel.vb
Imports System.Collections.ObjectModel
Imports System.Windows

Public Class ProductViewModel
  Inherits ViewModelBase
#Region "Constructor"
  Public Sub New()
    ' Make sure GetProducts has great exception handling!
    DataCollection = _DataManager.GetProducts()
  End Sub
#End Region

#Region "Private Variables"
  Private _DataManager As New ProductManager()
#End Region

#Region "DataCollection Property"
  Public Property DataCollection() As ObservableCollection(Of Product)
    Get
      Return DirectCast(GetValue(DataCollectionProperty), ObservableCollection(Of Product))
    End Get
    Set(ByVal value As ObservableCollection(Of Product))
      SetValue(DataCollectionProperty, value)
    End Set
  End Property

  Public Shared ReadOnly DataCollectionProperty As DependencyProperty = DependencyProperty.Register("DataCollection", GetType(ObservableCollection(Of Product)), GetType(ProductViewModel), New UIPropertyMetadata(Nothing))
#End Region

#Region "DetailData Property"
  Public Property DetailData() As Product
    Get
      Return DirectCast(GetValue(DetailDataProperty), Product)
    End Get
    Set(ByVal value As Product)
      SetValue(DetailDataProperty, value)
    End Set
  End Property

  Public Shared ReadOnly DetailDataProperty As DependencyProperty = DependencyProperty.Register("DetailData", GetType(Product), GetType(ProductViewModel), New UIPropertyMetadata(Nothing))
#End Region

#Region "AddRecord Method"
  Public Sub AddRecord()
    SetViewStateMode(UIStateMode.Add)

    ' Create Empty Object for UI to Display
    DetailData = New Product()
  End Sub
#End Region

#Region "CancelEdit Method"
  Public Sub CancelEdit()
    SetViewStateMode(UIStateMode.Cancel)

    ' TODO: Write Code to Undo Here
  End Sub
#End Region

#Region "SaveData Method"
  Public Sub SaveData()
    If IsAddMode Then
      _DataManager.Insert(DetailData)
      DataCollection.Add(DetailData)
      SetViewStateMode(UIStateMode.Normal)
    Else
      _DataManager.Update(DetailData)
      SetViewStateMode(UIStateMode.Normal)
    End If
  End Sub
#End Region

#Region "DeleteData Method"
  Public Sub DeleteData()
    _DataManager.Delete(DetailData)
    DataCollection.Remove(DetailData)
    SetViewStateMode(UIStateMode.Normal)
  End Sub
#End Region
End Class

File: _wpf/wpf-business-programmer/3-wpf-business-programmer-m3-exercise-files/Samples-2012/CSharp/WPFTemplateApp/WPFTemplateApp/ViewModels/ProductViewModel.cs
using System.Collections.ObjectModel;
using System.Windows;

namespace WPFTemplateApp
{
  public class ProductViewModel : ViewModelBase
  {
    #region Constructor
    public ProductViewModel()
    {
      // Make sure GetProducts has great exception handling!
      DataCollection = _DataManager.GetProducts();
    }
    #endregion

    #region Private Variables
    ProductManager _DataManager = new ProductManager();
    #endregion

    #region DataCollection Property
    public ObservableCollection<Product> DataCollection
    {
      get { return (ObservableCollection<Product>)GetValue(DataCollectionProperty); }
      set { SetValue(DataCollectionProperty, value); }
    }

    public static readonly DependencyProperty DataCollectionProperty =
        DependencyProperty.Register("DataCollection", typeof(ObservableCollection<Product>), typeof(ProductViewModel), new UIPropertyMetadata(null));
    #endregion
    
    #region DetailData Property
    public Product DetailData
    {
      get { return (Product)GetValue(DetailDataProperty); }
      set { SetValue(DetailDataProperty, value); }
    }

    public static readonly DependencyProperty DetailDataProperty =
        DependencyProperty.Register("DetailData", typeof(Product), typeof(ProductViewModel), new UIPropertyMetadata(null));
    #endregion

    #region AddRecord Method
    public void AddRecord()
    {
      SetViewStateMode(UIStateMode.Add);

      // Create Empty Object for UI to Display
      DetailData = new Product();
    }
    #endregion

    #region CancelEdit Method
    public void CancelEdit()
    {
      SetViewStateMode(UIStateMode.Cancel);

      // TODO: Write Code to Undo Here
    }
    #endregion

    #region SaveData Method
    public void SaveData()
    {
      if (IsAddMode)
      {
        _DataManager.Insert(DetailData);
        DataCollection.Add(DetailData);
        SetViewStateMode(UIStateMode.Normal);
      }
      else
      {
        _DataManager.Update(DetailData);
        SetViewStateMode(UIStateMode.Normal);
      }
    }
    #endregion

    #region DeleteData Method
    public void DeleteData()
    {
      _DataManager.Delete(DetailData);
      DataCollection.Remove(DetailData);
      SetViewStateMode(UIStateMode.Normal);
    }
    #endregion
  }
}
